// @asterkit/core/dist/index.js
// src/config.ts
var DEFAULT_HOST = "https://fapi.asterdex.com";
var DEFAULT_SIGNATURE_CHAIN_ID = 56;
var DEFAULT_AGENT_NAME = "AsterKit";
var DEFAULT_ASTER_CHAIN = "Mainnet";
var EIP712_DOMAIN_MAIN = {
  name: "AsterSignTransaction",
  version: "1",
  chainId: 56,
  verifyingContract: "0x0000000000000000000000000000000000000000"
};
var EIP712_DOMAIN_MESSAGE = {
  name: "AsterSignTransaction",
  version: "1",
  chainId: 1666,
  verifyingContract: "0x0000000000000000000000000000000000000000"
};

// src/utils.ts
var lastSecond = 0;
var nonceCounter = 0;
function getNonce() {
  const nowSecond = Math.floor(Date.now() / 1e3);
  if (nowSecond === lastSecond) {
    nonceCounter += 1;
  } else {
    lastSecond = nowSecond;
    nonceCounter = 0;
  }
  return nowSecond * 1e6 + nonceCounter;
}
function buildQueryString(params) {
  return Object.keys(params).filter((key) => params[key] !== void 0 && params[key] !== null).map((key) => `${key}=${String(params[key])}`).join("&");
}
function capitalizeKeys(obj) {
  const result = {};
  for (const [key, value] of Object.entries(obj)) {
    const capitalizedKey = key.charAt(0).toUpperCase() + key.slice(1);
    result[capitalizedKey] = value;
  }
  return result;
}
function inferMessageFieldType(value) {
  if (typeof value === "boolean") return "bool";
  if (typeof value === "bigint") return "uint256";
  if (typeof value === "number" && Number.isInteger(value)) return "uint256";
  return "string";
}
function resolveSignerAccount(walletClient) {
  const selected = walletClient.account;
  if (selected === void 0) {
    throw new Error(
      "walletClient.account is required. Create walletClient with an account."
    );
  }
  return selected;
}
async function signEIP712Main({
  walletClient,
  params,
  primaryType
}) {
  const signerAccount = resolveSignerAccount(walletClient);
  const capitalizedParams = capitalizeKeys(params);
  const messageType = Object.entries(
    capitalizedParams
  ).map(([name, value]) => ({
    name,
    type: inferMessageFieldType(value)
  }));
  const types = {
    [primaryType]: messageType
  };
  const typedDataParams = {
    account: signerAccount,
    domain: EIP712_DOMAIN_MAIN,
    primaryType,
    types,
    message: capitalizedParams
  };
  return walletClient.signTypedData(typedDataParams);
}
async function signEIP712Message({
  walletClient,
  message
}) {
  const signerAccount = resolveSignerAccount(walletClient);
  return walletClient.signTypedData({
    account: signerAccount,
    domain: EIP712_DOMAIN_MESSAGE,
    primaryType: "Message",
    types: {
      Message: [{ name: "msg", type: "string" }]
    },
    message: {
      msg: message
    }
  });
}

// src/agent.ts
var AsterRequestError = class extends Error {
  status;
  data;
  url;
  constructor(status, url, data) {
    super(`Aster request failed with status ${status}`);
    this.name = "AsterRequestError";
    this.status = status;
    this.url = url;
    this.data = data;
  }
};
async function approveAgent({
  walletClient,
  host,
  signatureChainId,
  agentName,
  agentAddress,
  ipWhitelist,
  expired,
  canSpotTrade,
  canPerpTrade,
  canWithdraw,
  builder,
  maxFeeRate,
  builderName,
  asterChain,
  nonce
}) {
  const defaultExpired = (() => {
    const nextMonth = /* @__PURE__ */ new Date();
    nextMonth.setMonth(nextMonth.getMonth() + 1);
    return nextMonth.getTime();
  })();
  const walletUser = walletClient.account?.address;
  if (!walletUser) {
    throw new Error(
      "walletClient.account is required. Create walletClient with an account."
    );
  }
  const normalizedIpWhitelist = ipWhitelist?.trim();
  const params = {
    agentName: agentName ?? DEFAULT_AGENT_NAME,
    agentAddress,
    ...normalizedIpWhitelist ? { ipWhitelist: normalizedIpWhitelist } : {},
    expired: expired ?? defaultExpired,
    canSpotTrade: canSpotTrade ?? true,
    canPerpTrade: canPerpTrade ?? true,
    canWithdraw: canWithdraw ?? false,
    ...builder ? { builder } : {},
    ...maxFeeRate ? { maxFeeRate } : {},
    ...builderName ? { builderName } : {},
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user: walletUser,
    nonce: nonce ?? getNonce()
  };
  const signature = await signEIP712Main({
    walletClient,
    params,
    primaryType: "ApproveAgent"
  });
  const signedParams = {
    ...params,
    signature,
    signatureChainId: signatureChainId ?? DEFAULT_SIGNATURE_CHAIN_ID
  };
  const queryString = buildQueryString(signedParams);
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/approveAgent?${queryString}`;
  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    },
    body: ""
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: signedParams
  };
}
async function getAgents({
  walletClient,
  host,
  asterChain,
  user,
  signer,
  nonce,
  signature
}) {
  const params = {
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user,
    signer,
    nonce: nonce ?? getNonce()
  };
  const queryString = buildQueryString(params);
  let finalSignature = signature;
  if (!finalSignature) {
    if (!walletClient) {
      throw new Error(
        "getAgents requires either signature or walletClient to sign the query."
      );
    }
    if (!walletClient.account?.address) {
      throw new Error(
        "walletClient.account is required. Create walletClient with an account."
      );
    }
    if (walletClient.account.address.toLowerCase() !== signer.toLowerCase()) {
      throw new Error("signer must match walletClient.account.address.");
    }
    finalSignature = await signEIP712Message({
      walletClient,
      message: queryString
    });
  }
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/agent?${queryString}&signature=${finalSignature}`;
  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    }
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: {
      ...params,
      signature: finalSignature
    }
  };
}
async function updateAgent({
  walletClient,
  host,
  signatureChainId,
  agentAddress,
  ipWhitelist,
  canSpotTrade,
  canPerpTrade,
  canWithdraw,
  asterChain,
  nonce
}) {
  const walletUser = walletClient.account?.address;
  if (!walletUser) {
    throw new Error(
      "walletClient.account is required. Create walletClient with an account."
    );
  }
  const params = {
    agentAddress,
    canSpotTrade,
    canPerpTrade,
    canWithdraw,
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user: walletUser,
    nonce: nonce ?? getNonce()
  };
  if (ipWhitelist !== void 0 && ipWhitelist !== null) {
    params.ipWhitelist = ipWhitelist;
  }
  const signature = await signEIP712Main({
    walletClient,
    params,
    primaryType: "UpdateAgent"
  });
  const signedParams = {
    ...params,
    signature,
    signatureChainId: signatureChainId ?? DEFAULT_SIGNATURE_CHAIN_ID
  };
  const queryString = buildQueryString(signedParams);
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/updateAgent?${queryString}`;
  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    },
    body: ""
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: signedParams
  };
}
async function deleteAgent({
  walletClient,
  host,
  signatureChainId,
  agentAddress,
  asterChain,
  nonce
}) {
  const walletUser = walletClient.account?.address;
  if (!walletUser) {
    throw new Error(
      "walletClient.account is required. Create walletClient with an account."
    );
  }
  const params = {
    agentAddress,
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user: walletUser,
    nonce: nonce ?? getNonce()
  };
  const signature = await signEIP712Main({
    walletClient,
    params,
    primaryType: "DelAgent"
  });
  const signedParams = {
    ...params,
    signature,
    signatureChainId: signatureChainId ?? DEFAULT_SIGNATURE_CHAIN_ID
  };
  const queryString = buildQueryString(signedParams);
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/agent?${queryString}`;
  const response = await fetch(url, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    }
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: signedParams
  };
}

// src/builder.ts
async function approveBuilder({
  walletClient,
  host,
  signatureChainId,
  builder,
  maxFeeRate,
  builderName,
  asterChain,
  nonce
}) {
  const walletUser = walletClient.account?.address;
  if (!walletUser) {
    throw new Error(
      "walletClient.account is required. Create walletClient with an account."
    );
  }
  const params = {
    builder,
    maxFeeRate,
    builderName,
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user: walletUser,
    nonce: nonce ?? getNonce()
  };
  const signature = await signEIP712Main({
    walletClient,
    params,
    primaryType: "ApproveBuilder"
  });
  const signedParams = {
    ...params,
    signature,
    signatureChainId: signatureChainId ?? DEFAULT_SIGNATURE_CHAIN_ID
  };
  const queryString = buildQueryString(signedParams);
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/approveBuilder?${queryString}`;
  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    },
    body: ""
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: signedParams
  };
}
async function updateBuilder({
  walletClient,
  host,
  signatureChainId,
  builder,
  maxFeeRate,
  asterChain,
  nonce
}) {
  const walletUser = walletClient.account?.address;
  if (!walletUser) {
    throw new Error(
      "walletClient.account is required. Create walletClient with an account."
    );
  }
  const params = {
    builder,
    maxFeeRate,
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user: walletUser,
    nonce: nonce ?? getNonce()
  };
  const signature = await signEIP712Main({
    walletClient,
    params,
    primaryType: "UpdateBuilder"
  });
  const signedParams = {
    ...params,
    signature,
    signatureChainId: signatureChainId ?? DEFAULT_SIGNATURE_CHAIN_ID
  };
  const queryString = buildQueryString(signedParams);
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/updateBuilder?${queryString}`;
  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    },
    body: ""
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: signedParams
  };
}
async function getBuilders({
  walletClient,
  host,
  asterChain,
  user,
  signer,
  nonce,
  signature
}) {
  const params = {
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user,
    signer,
    nonce: nonce ?? getNonce()
  };
  const queryString = buildQueryString(params);
  let finalSignature = signature;
  if (!finalSignature) {
    if (!walletClient) {
      throw new Error(
        "getBuilders requires either signature or walletClient to sign the query."
      );
    }
    if (!walletClient.account?.address) {
      throw new Error(
        "walletClient.account is required. Create walletClient with an account."
      );
    }
    if (walletClient.account.address.toLowerCase() !== signer.toLowerCase()) {
      throw new Error("signer must match walletClient.account.address.");
    }
    finalSignature = await signEIP712Message({
      walletClient,
      message: queryString
    });
  }
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/builder?${queryString}&signature=${finalSignature}`;
  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    }
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: {
      ...params,
      signature: finalSignature
    }
  };
}
async function deleteBuilder({
  walletClient,
  host,
  signatureChainId,
  builder,
  asterChain,
  nonce
}) {
  const walletUser = walletClient.account?.address;
  if (!walletUser) {
    throw new Error(
      "walletClient.account is required. Create walletClient with an account."
    );
  }
  const params = {
    builder,
    asterChain: asterChain ?? DEFAULT_ASTER_CHAIN,
    user: walletUser,
    nonce: nonce ?? getNonce()
  };
  const signature = await signEIP712Main({
    walletClient,
    params,
    primaryType: "DelBuilder"
  });
  const signedParams = {
    ...params,
    signature,
    signatureChainId: signatureChainId ?? DEFAULT_SIGNATURE_CHAIN_ID
  };
  const queryString = buildQueryString(signedParams);
  const url = `${host ?? DEFAULT_HOST}/fapi/v3/builder?${queryString}`;
  const response = await fetch(url, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": "AsterKit/1.0"
    }
  });
  const data = await response.json();
  if (!response.ok) {
    throw new AsterRequestError(response.status, url, data);
  }
  return {
    status: response.status,
    data,
    url,
    params: signedParams
  };
}
export {
  AsterRequestError,
  DEFAULT_AGENT_NAME,
  DEFAULT_ASTER_CHAIN,
  DEFAULT_HOST,
  DEFAULT_SIGNATURE_CHAIN_ID,
  EIP712_DOMAIN_MAIN,
  EIP712_DOMAIN_MESSAGE,
  approveAgent,
  approveBuilder,
  buildQueryString,
  capitalizeKeys,
  deleteAgent,
  deleteBuilder,
  getAgents,
  getBuilders,
  getNonce,
  signEIP712Main,
  signEIP712Message,
  updateAgent,
  updateBuilder
};
